File: typescript/triangle/triangle.ts
export default class Triangle {

    sides: number[]

    constructor(...sides: number[]) {
        this.sides = sides
    }

    kind() {
        // validate sides
        if (this.sides.some(side => side <= 0)) {
            throw new Error('Invalid length')
        }
        const [a, b, c] = this.sides
        // check for inequality
        if (a + b < c || a + c < b || b + c < a) {
            throw new Error('Triangle inequality')
        }
        // all sides equal
        if (this.sides.every(side => side === a)) {
            return 'equilateral'
        }
        // any two sides equal
        if (a === b || a === c || b === c) {
            return 'isosceles'
        }
        return 'scalene'
    }
}

File: dart/triangle/lib/triangle.dart
class Triangle {
  bool equilateral(double a, double b, double c) => a != 0 && a == b && b == c;

  bool isosceles(double a, double b, double c) {
    return this.isValidTriangle(a, b, c) && (a == b || a == c || b == c);
  }

  bool scalene(double a, double b, double c) {
    return this.isValidTriangle(a, b, c) && a != b && a != c && b != c;
  }

  bool isValidTriangle(double a, double b, double c) {
    if (a == 0 || b == 0 || c == 0) {
      return false;
    }
    return !(a + b < c || a + c < b || b + c < a);
  }
}
